README.md
# MailLift notifications channel for Laravel 5.3

[![Latest Version on Packagist](https://img.shields.io/packagist/v/laravel-notification-channels/maillift.svg?style=flat-square)](https://packagist.org/packages/laravel-notification-channels/maillift)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/laravel-notification-channels/maillift/master.svg?style=flat-square)](https://travis-ci.org/laravel-notification-channels/maillift)
[![StyleCI](https://styleci.io/repos/66273929/shield)](https://styleci.io/repos/65743131)
[![SensioLabsInsight](https://img.shields.io/sensiolabs/i/483b38bc-e579-4661-8a97-a469e1e24988.svg?style=flat-square)](https://insight.sensiolabs.com/projects/483b38bc-e579-4661-8a97-a469e1e24988)
[![Quality Score](https://img.shields.io/scrutinizer/g/laravel-notification-channels/maillift.svg?style=flat-square)](https://scrutinizer-ci.com/g/laravel-notification-channels/maillift)
[![Code Coverage](https://img.shields.io/scrutinizer/coverage/g/laravel-notification-channels/maillift/master.svg?style=flat-square)](https://scrutinizer-ci.com/g/laravel-notification-channels/maillift/?branch=master)
[![Total Downloads](https://img.shields.io/packagist/dt/laravel-notification-channels/maillift.svg?style=flat-square)](https://packagist.org/packages/laravel-notification-channels/maillift)

This package makes it easy to create [MailLift tasks](https://maillift.com/) with Laravel 5.3.

## Contents

- [Installation](#installation)
    - [Setting up the MailLift service](#setting-up-the-maillift-service)
- [Usage](#usage)
	- [Available Message methods](#available-message-methods)
- [Changelog](#changelog)
- [Testing](#testing)
- [Security](#security)
- [Contributing](#contributing)
- [Credits](#credits)
- [License](#license)


## Installation

You can install the package via composer:

``` bash
composer require laravel-notification-channels/maillift
```

### Setting up the MailLift service

Register at [maillift.com](https://maillift.com).

Add your MailLift username and API key to your `config/services.php`:

```php
// config/services.php
...
'maillift' => [
    'user' => env('MAILLIFT_USERNAME'),
    'key' => env('MAILLIFT_API_KEY'),
],
...
```

## Usage

Now you can use the channel in your `via()` method inside the notification:

``` php
use NotificationChannels\MailLift\MailLiftChannel;
use NotificationChannels\MailLift\MailLiftMessage;
use Illuminate\Notifications\Notification;

class ProjectCreated extends Notification
{
    public function via($notifiable)
    {
        return [MailLiftChannel::class];
    }

    public function toMailLift($notifiable)
    {
        return MailLiftMessage::create('This is my handwritten letter body')
            ->sender('Laravel Notification Channels'. PHP_EOL . 'Some street 123');
    }
}
```

In order to let your notification know which address you want to send the handwritten letter to, add the `routeNotificationForMailLift` method to your Notifiable model.

This method needs to return a string with the recipient address. Use line breaks (\n character) to separate lines in the field.

```php
public function routeNotificationForMailLift()
{
    return 'Recipient Name' . PHP_EOL . 'Recipient Address' . PHP_EOL . 'Recipient State / Postal Code';
}
```

### Available methods

- `body('')`: Accepts a string value for the MailLift letter body.
- `sender('')`: Accepts a string value for the MailLift sender. Use line breaks (\n character) to separate lines in the field.
- `scheduleDelivery('')`: Accepts a string or DateTime object for the scheduled delivery date.


## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

## Security

If you discover any security related issues, please email m.pociot@gmail.com instead of using the issue tracker.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Credits

- [Marcel Pociot](https://github.com/mpociot)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `maillift` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/ChannelTest.php
<?php

namespace NotificationChannels\MailLift\Test;

use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Response;
use Illuminate\Notifications\Notification;
use Mockery;
use NotificationChannels\MailLift\Exceptions\CouldNotSendNotification;
use NotificationChannels\MailLift\Exceptions\InvalidConfiguration;
use NotificationChannels\MailLift\MailLiftChannel;
use NotificationChannels\MailLift\MailLiftMessage;
use Orchestra\Testbench\TestCase;

class ChannelTest extends TestCase
{
    /** @test */
    public function it_can_send_a_notification()
    {
        $this->app['config']->set('services.maillift.user', 'MailLiftUsername');
        $this->app['config']->set('services.maillift.key', 'MailLiftKey');
        $response = new Response(200);
        $client = Mockery::mock(Client::class);
        $client->shouldReceive('post')
            ->once()
            ->with('https://api.maillift.com/2016-04-21/letter/', Mockery::type('array'))
            ->andReturn($response);
        $channel = new MailLiftChannel($client);
        $channel->send(new TestNotifiable(), new TestNotification());
    }

    /** @test */
    public function it_throws_an_exception_when_it_is_not_configured()
    {
        $this->setExpectedException(InvalidConfiguration::class);

        $client = new Client();
        $channel = new MailLiftChannel($client);
        $channel->send(new TestNotifiable(), new TestNotification());
    }

    /** @test */
    public function it_throws_an_exception_when_it_could_not_send_the_notification()
    {
        $this->app['config']->set('services.maillift.user', 'MailLiftUsername');
        $this->app['config']->set('services.maillift.key', 'MailLiftKey');

        $response = new Response(500);
        $client = Mockery::mock(Client::class);
        $client->shouldReceive('post')
            ->once()
            ->andReturn($response);
        $channel = new MailLiftChannel($client);

        $this->setExpectedException(CouldNotSendNotification::class);

        $channel->send(new TestNotifiable(), new TestNotification());
    }
}

class TestNotifiable
{
    use \Illuminate\Notifications\Notifiable;

    /**
     * @return int
     */
    public function routeNotificationForMailLift()
    {
        return 'Laravel Notification Channels'.PHP_EOL.'Receiver Street 123';
    }
}


class TestNotification extends Notification
{
    public function toMailLift($notifiable)
    {
        return MailLiftMessage::create('Message body')
            ->sender('Laravel Notification Channels'.PHP_EOL.'Street 1');
    }
}


tests/MessageTest.php
<?php

namespace NotificationChannels\MailLift\Test;

use DateTime;
use Illuminate\Support\Arr;
use NotificationChannels\MailLift\MailLiftMessage;

class MessageTest extends \PHPUnit_Framework_TestCase
{
    /** @var \NotificationChannels\MailLift\MailLiftMessage */
    protected $message;

    public function setUp()
    {
        parent::setUp();

        $this->message = new MailLiftMessage('');
    }

    /** @test */
    public function it_accepts_a_body_when_constructing_a_message()
    {
        $message = new MailLiftMessage('MailBody');

        $this->assertEquals('MailBody', Arr::get($message->toArray(), 'MessageBody'));
    }

    /** @test */
    public function it_provides_a_create_method()
    {
        $message = MailLiftMessage::create('MailBody');

        $this->assertEquals('MailBody', Arr::get($message->toArray(), 'MessageBody'));
    }

    /** @test */
    public function it_can_set_the_body()
    {
        $this->message->body('MailBody');

        $this->assertEquals('MailBody', Arr::get($this->message->toArray(), 'MessageBody'));
    }

    /** @test */
    public function it_can_set_the_sender()
    {
        $this->message->sender('MailSender');

        $this->assertEquals('MailSender', Arr::get($this->message->toArray(), 'Sender'));
    }

    /** @test */
    public function it_can_set_a_scheduled_date_from_string()
    {
        $date = new DateTime('tomorrow');
        $this->message->scheduleDelivery('tomorrow');

        $this->assertEquals($date->format('Y-m-d'), Arr::get($this->message->toArray(), 'ScheduledDelivery'));
    }

    /** @test */
    public function it_can_set_a_scheduled_date_from_datetime()
    {
        $date = new DateTime('tomorrow');
        $this->message->scheduleDelivery($date);

        $this->assertEquals($date->format('Y-m-d'), Arr::get($this->message->toArray(), 'ScheduledDelivery'));
    }
}


src/MailLiftMessage.php
<?php

namespace NotificationChannels\MailLift;

use DateTime;

class MailLiftMessage
{
    /** @var string */
    protected $body;

    /** @var string */
    protected $sender;

    /** @var string|null */
    protected $scheduledDelivery;

    /**
     * @param string $body
     *
     * @return static
     */
    public static function create($body)
    {
        return new static($body);
    }

    /**
     * @param string $body
     */
    public function __construct($body)
    {
        $this->body = $body;
    }

    /**
     * Set the letter message body.
     *
     * @param $body
     *
     * @return $this
     */
    public function body($body)
    {
        $this->body = $body;

        return $this;
    }

    /**
     * Set the letter message sender.
     *
     * @param $sender
     *
     * @return $this
     */
    public function sender($sender)
    {
        $this->sender = $sender;

        return $this;
    }

    /**
     * Set the scheduled delivery date.
     *
     * @param string|DateTime $scheduledDelivery
     *
     * @return $this
     */
    public function scheduleDelivery($scheduledDelivery)
    {
        if (! $scheduledDelivery instanceof DateTime) {
            $scheduledDelivery = new DateTime($scheduledDelivery);
        }

        $this->scheduledDelivery = $scheduledDelivery->format('Y-m-d');

        return $this;
    }

    /**
     * @return array
     */
    public function toArray()
    {
        return [
            'MessageBody' => $this->body,
            'Sender' => $this->sender,
            'ScheduledDelivery' => $this->scheduledDelivery,
        ];
    }
}


src/MailLiftChannel.php
<?php

namespace NotificationChannels\MailLift;

use GuzzleHttp\Client;
use Illuminate\Support\Arr;
use NotificationChannels\MailLift\Exceptions\CouldNotSendNotification;
use Illuminate\Notifications\Notification;
use NotificationChannels\MailLift\Exceptions\InvalidConfiguration;

class MailLiftChannel
{
    const API_ENDPOINT = 'https://api.maillift.com/2016-04-21/letter/';

    /** @var Client */
    protected $client;

    /**
     * @param Client $client
     */
    public function __construct(Client $client)
    {
        $this->client = $client;
    }

    /**
     * Send the given notification.
     *
     * @param mixed $notifiable
     * @param \Illuminate\Notifications\Notification $notification
     *
     * @throws \NotificationChannels\MailLift\Exceptions\InvalidConfiguration
     * @throws \NotificationChannels\MailLift\Exceptions\CouldNotSendNotification
     */
    public function send($notifiable, Notification $notification)
    {
        if (! $routing = $notifiable->routeNotificationFor('MailLift')) {
            return;
        }

        $user = config('services.maillift.user');
        $key = config('services.maillift.key');

        if (is_null($key) || is_null($user)) {
            throw InvalidConfiguration::configurationNotSet();
        }

        $mailliftParameters = $notification->toMailLift($notifiable)->toArray();

        $response = $this->client->post(self::API_ENDPOINT, [
            'auth' => [$user, $key],
            'form_params' => Arr::set($mailliftParameters, 'Recipient', $routing),
        ]);

        if ($response->getStatusCode() !== 200) {
            throw CouldNotSendNotification::serviceRespondedWithAnError($response);
        }
    }
}
